### ISSUE TITLE
CPU100%,When I add the "evhttp_connection_free" in the callback function "http_request_done"

### ISSUE THREAD

**Issue by moonwalkerzyp** [libevent/libevent#560]
I make a demo like this: in the main process, i create the event in the event_base listen to the client request,then make a http request to the httpserver, and use the "evhttp_connection_free" to clean the connection, then i fork a process do nothing,just sleep ,when   multi client concurrent connect to me, the cpu100%, strace the main process id ,i found lots of "EPOLLIN|EPOLLERR|EPOLLHUP"  events from epoll_wait,the cpu is nomal if the client connection interval change to the 1s。

here is my demo:
```
void http_request_done(struct evhttp_request *req, void *arg)  
{  
    pid_t pid;
    struct evhttp_connection *conn = (struct evhttp_connection*)arg;
    size_t len = evbuffer_get_length(req->input_buffer);  
    unsigned char * str = evbuffer_pullup(req->input_buffer, len);  
    char buf[1024] = { 0 };  
    memcpy(buf, str, len);  
    if(!strcmp(buf, "+OK")){
        evhttp_connection_free(conn);  //**the cpu is normal if i not free the connection]**
        pid = fork();   // the cpu is normal if i did not fork a process
        if(pid == 0){
            while(1){
                sleep(1);
            }
    }
}

void main_callback(int sock, short event, void* arg){

    struct event_base *base = (struct event_base*)arg;
    struct evhttp_connection* conn;  
    struct evhttp_request* req;  
    int iret = 0;
    int buf_len = 0;
    int address_size = sizeof(struct sockaddr_in);
    struct sockaddr_in rin;
    char buf[1024] = {0};
    iret = recvfrom(sock, buf, 1024 - 1, 0, (struct sockaddr *)&rin, &address_size);
    if (iret < 0 || iret >= 1024 - 1) {
        return;
    }
    buf[iret] = '\0';
    printf("recv buf:%s,buflen:%d\n",buf,iret);
    buf_len = iret;
    if(!strcmp(buf, "login")){
        conn = evhttp_connection_new("httpserver ip", http port);  
        evhttp_connection_set_base(conn, base);  
        req = evhttp_request_new(http_request_done, conn);  
        evhttp_add_header(req->output_headers, "Host", "localhost");  
        evhttp_make_request(conn, req, EVHTTP_REQ_GET, "/index.php?id=1");  
        evhttp_connection_set_timeout(req->evcon, 1);  
    }
} 

int main()  
{  
    struct event_base* base;  
    struct event listen_ev;
    int udp_fd;
    int ret = 0;

    base = event_base_new();  

    udp_fd = rs_udp_socket_create("server ip", port);
    if (udp_fd < 0) {
        printf("%s\n","socket create failed!");
    }
    ret = rs_socket_nonblock_set(udp_fd);
    if(ret!=0){
        printf("set nonblock failed\n");
    }

    event_set(&listen_ev, udp_fd, EV_READ | EV_PERSIST, main_callback , (void *)base);
    ret = event_base_set(base, &listen_ev);
    if (ret < 0) {
        printf("event set failed\n");
        return -1;
    }
    event_add(&listen_ev, NULL);
    event_base_dispatch(base);  
    event_base_free(base);  
    printf("run over...\n");  
    return 0;  
}  

```

**Comment by azat** [comment]
Hi

>         pid = fork();   // the cpu is normal if i did not fork a process

Fork is not a good solution here I guess, but anyway if you use it,
you may consider looking at event_reinit().
But this is not about your question.

>     iret = recvfrom(sock, buf, 1024 - 1, 0, (struct sockaddr *)&rin, &address_size);

I guess that rs_udp_socket_create() creates socket with O_CLOEXEC/SOCK_CLOEXEC?

Azat.

**Comment by azat** [comment]
On Thu, Sep 21, 2017 at 10:11 PM, Azat Khuzhin <a3at.mail@gmail.com> wrote:
> Hi
>
>>         pid = fork();   // the cpu is normal if i did not fork a process
>
> Fork is not a good solution here I guess, but anyway if you use it,
> you may consider looking at event_reinit().
> But this is not about your question.
>
>>     iret = recvfrom(sock, buf, 1024 - 1, 0, (struct sockaddr *)&rin, &address_size);
>
> I guess that rs_udp_socket_create() creates socket with O_CLOEXEC/SOCK_CLOEXEC?

Ignore my question, I misread something.
Will get back soon..

**Comment by azat** [comment]
>         conn = evhttp_connection_new("httpserver ip", http port);

Actually your program should fail here (previous line), because you
are mixing deprecated and new API.
evhttp_connection_new() will use *global* event_base, which installed
only in event_init(), you should use evhttp_connection_base_new()
instead.

>         evhttp_connection_set_base(conn, base);

And this doesn't changes anything, since evhttp_connection_new() needs
event_base before the above call.

>     base = event_base_new();

This doesn't install global event base.

You can take a look at your modified sample at:
  https://gist.github.com/azat/6d29ddd3ec743294ed1469303f5d3bc5

Also:
- event_set is deprecated, event_new()/event_assign() should be used instead
- using evhttp_request structure should be replaced with getters

Basically, the rule of thumb, if there is a function that has/accepts
"base" in it's name it should be used instead.
And do not use "_compat.h"/"_struct.h" headers if you don't need them
explicitly.

Hope this helps,
Azat.

**Comment by moonwalkerzyp** [comment]
thank you for your help, i try your modified sample, and i try to kill a signal fork in other place ,buf the question still exits,
here is my strace process id result when the cpu runs 100%:
 [strace.txt](https://github.com/libevent/libevent/files/1323491/strace.txt)

and client、server sample at: 
 https://gist.github.com/moonwalkerzyp/d1f22b168736418c85981b9970081a27

**Comment by azat** [comment]
> and i try to kill a signal fork in other place

This shouldn't change anything.

Which version of libevent do you use?

**Comment by moonwalkerzyp** [comment]
Thank you for your reminding，I change my libevent version libevent-2.0.22-stable  to libevent-2.1.18-stable, the problem is solved！ 
and Can you explain briefly the nature of the problem，i feel puzzle about the EPOLLIN|EPOLLERR|EPOLLHUP。
think you for your time。

**Comment by azat** [comment]
> my libevent version libevent-2.0.22-stable

Hm, I cannot reproduce your problem under latest 2.0 (my latest 2.0 is
11 patches ahead of 2.0.22 release, but I don't think that they
changes anything).

> and Can you explain briefly the nature of the problem，i feel puzzle about the EPOLLIN|EPOLLERR|EPOLLHUP。

Well as you can figure out from man this errors means that peer closed
the connection, and if we will not remove it from event_base we will
always get this events.

Maybe you have another thread runs that can change something? Also
does your strace.log is full (i.e. without filters)? Can you attach
full (basically start program under strace, not attach after) strace
log (strace -f -tt -s100) when you can reproduce this issue?

The relevant lines from strace is:
  socket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 8
  ...
  fcntl(8, F_SETFD, FD_CLOEXEC)           = 0 <-- this means that we
will not have it in new childs
  ...
  close(8)                                = 0 <-- so because of
FD_CLOEXEC it should be automatically removed from epoll
  ..
  epoll_ctl(3, EPOLL_CTL_DEL, 8, {EPOLLIN, {u32=8, u64=8}}) = -1 EBADF
(Bad file descriptor) <-- and indeed, it already removed
  epoll_wait(3, {{EPOLLIN|EPOLLERR|EPOLLHUP, {u32=8, u64=8}},
{EPOLLIN, {u32=6, u64=6}}}, 32, -1) = 2 <-- but why fd 8 is here?

Also can you send me:
- glibc version
- linux kernel version (uname -a)

Another strange thing to me is that you have SO_KEEPALIVE (and others)
options installed, though your sample doesn't use functions that will
do this (from libevent).

And one more thing for your sources, you should check if "req" is
valid in request callback (as you can see in mine I had just assert()
for simplicity reasons), but even if it will be NULL it shouldn't
break things like this.

**Comment by moonwalkerzyp** [comment]
> Maybe you have another thread runs that can change something? Also
> does your strace.log is full (i.e. without filters)? Can you attach
> full (basically start program under strace, not attach after) strace
> log (strace -f -tt -s100) when you can reproduce this issue?

just like the sample，there is no another thread runs, and i reproduce the full strace.log, 
[strace_full_libevent2.0.22.log](https://github.com/libevent/libevent/files/1326286/strace_full_libevent2.0.22.log)
and i also produce a normal strace.log under the libevent2.1.8 , mybe you can make a compare if you need,
[strace_full_libevent2.1.8.log](https://github.com/libevent/libevent/files/1326290/strace_full_libevent2.1.8.log)

> Also can you send me:
> - glibc version
> - linux kernel version (uname -a)

my glibc version : (Ubuntu EGLIBC 2.19-0ubuntu6.5) 2.19
kernel version:Linux ubuntu 3.16.0-30-generic #40~14.04.1-Ubuntu SMP Thu Jan 15 17:43:14 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

> SO_KEEPALIVE (and others)
> options installed, though your sample doesn't use functions that will
> do this (from libevent).

i did not see this log in the strace.log under the libevent2.1.8, is only shown when my sample runs under the libevent2.0.22,if the functions from libevent I use not called this, I really can't think of any other reason.

Thank you for your reply,
ZYP.

**Comment by azat** [comment]
> SO_KEEPALIVE (and others)
> options installed, though your sample doesn't use functions that will
> do this (from libevent).

SO_KEEPALIVE is used if you are using
`evhttp_connection_set_local_address()` but I didn't saw it there.
Can you also attach full `ltrace -f` log for 2.0?

**Comment by moonwalkerzyp** [comment]
> Can you also attach full `ltrace -f` log for 2.0?
ok，but i didn't see the `evhttp_connection_set_local_address()` in it
[ltrace2.0.log](https://github.com/libevent/libevent/files/1327396/ltrace2.0.log)

**Comment by azat** [comment]
> Can you also attach full ltrace -f log for 2.0?
> ok，but i didn't see the evhttp_connection_set_local_address() in it
> ltrace2.0.log

Ops, I missed [1], and it changes the behavior of when SO_KEEPALIVE
will be installed (and I think that we should always set it, otherwise
it is too tricky).

But even after checking out 2.0.22-stable (without any patches ahead,
including mentioned), I still cannot reproduce issue with my gist
(updated just now):
  https://gist.github.com/azat/6d29ddd3ec743294ed1469303f5d3bc5

[1] 502c043a3aadabba8ce85db2832654676811bf8b ("Fix ipv6 support for
http. When URL contain domain, not IP address.")

**Comment by moonwalkerzyp** [comment]
>  it changes the behavior of when SO_KEEPALIVE
> will be installed   

  what's that means? 
  Hm I make campare about the version 2.0.22 and version 2.1.8 of the libevent(source code)，I found the SO_KEEP_ALIVE is both set in "bind_socket_ai()" which is called by the function "evhttp_connection_connect", i feel strange why the 2.1.8 strace.log didn't show it,(beacuse of the patches you mentioned ?)

and also,another problem what we care is.

> epoll_ctl(3, EPOLL_CTL_DEL, 8, {EPOLLIN, {u32=8, u64=8}}) = -1 EBADF
> (Bad file descriptor) <-- and indeed, it already removed
>   epoll_wait(3, {{EPOLLIN|EPOLLERR|EPOLLHUP, {u32=8, u64=8}},
> {EPOLLIN, {u32=6, u64=6}}}, 32, -1) = 2 <-- but why fd 8 is here?

 Does the ”SO_KEEP_ALIVE“ is Associated with this problem？

**Comment by azat** [comment]
>   Hm I make campare about the version 2.0.22 and version 2.1.8 of the libevent(source code)，I found the SO_KEEP_ALIVE is both set in "bind_socket_ai()" which is called by the function "evhttp_connection_connect", i feel strange why the 2.1.8 strace.log didn't show it,(beacuse of the patches you mentioned ?)

Can you do a bisect?

>  Does the ”SO_KEEP_ALIVE“ is Associated with this problem？

No, at least I don't think so.

**Comment by azat** [comment]
> Can you do a bisect?

Forget it, I got it.

So the patch that fixes this issue is -- ab3bc69f47925299a919493f306b9d6fbb805b00
("http: avoid epoll_ctl() on already closed fd (triggers by http/chunk_out)")

But what happens here, why you need evhttp_connection_free() *and*
fork() here (I will talk about behaviour without mentioned patch) and
why it takes so much time to dig:

evhttp_connection_free() does this:
  close(fd)
  epoll_ctl(epoll_fd, DEL, fd)

fork() will inherit all fds that you had before (and what confused me
was then thing that I saw O_CLOEXEC flags, but by some strange reason I
though that execve() also was there, but it wasn't hence all fds will be
in childs until you call exec())

And when you sending requests to the server asynchronously you can have
next chain:
  main_callback() # fd=7
  main_callback() # fd=8
  http_request_done() # for fd=8
    fork() # will have fd=7, IOW now you have copy of fd=7
           # and this means that epoll will not remove this fd
           # automatically because it removes fd from set only when
           # there are now other references (see man epoll for more)

And this is why you need fork()

(It was a problem for me to reproduce this bug, because I had http
server which replies was very fast, and I can't get into this situation).

As for why you need evhttp_connection_free(), because it closes fd, and
hence you will get tons of EPOLLERR events for that fd (that has a
reference in child).

And what about patch, it firstly remove event and only after closes fd,
so you will not have EPOLLERR because event removed. But if you/libevent
don't remove event manually before close it will not able to remove it
later because in this process/thread there is no such fd already.

**Comment by moonwalkerzyp** [comment]
Many thanks for your help ,I'm totally understand,I close this issue.
